backend/apps/readings/analysis.py

```diff
@@ -33,6 +34,100 @@ def total_view_time(self):
+            self.mean_reading_time_for_a_question(question_three, context_one),
+            self.mean_reading_time_for_a_question(question_one, context_two),
+            self.mean_reading_time_for_a_question(question_two, context_two),
+            self.mean_reading_time_for_a_question(question_three, context_two)
+        ]
+
+        return mean_reading_time_results_data
+
+    def mean_reading_time_for_a_question(self, question, context):
+        """
+        Given the student response dicts, computes the mean read time for a
+        specific question (given by its keyword) and the context in which it was asked.
+        Returns the question, context, mean read time, and number of people who read.
+        :param question: string, to determine which question was being asked
+        :param context: string, what the reader thought the reading was
+        :return: tuple, in order of the question asked (full question), the context,
+        the mean read time, and the number of people who read it
+        """
+        mean_time = 0
+        number_of_readers = 0
+        question_count = 0
+        reading_time = []
+        total_question_view_time = 0
+        student_data = self.responses[:]
+        for response in student_data:
+            if question != response.question.text or \
+               context != response.context.text:
+                continue
+            if len(response.get_parsed_views()) != 0:
+                number_of_readers += 1
+            for view_time in response.get_parsed_views():
+                reading_time.append(view_time)
+
+        if len(reading_time) != 0:
+            self.remove_outliers(reading_time)
+
+        view_time = 0
+        while view_time < len(reading_time):
+            question_count += 1
+            total_question_view_time += reading_time[view_time]
+            view_time += 1
+
+        if len(reading_time) != 0:
+            mean_time = round(total_question_view_time / len(reading_time), 2)
+
+        return "Question: " + question + " Context: "+context+" Mean time with outliers " \
+            "removed: " + str(mean_time) + " Total number of readers: " + str(number_of_readers)
+
+    def remove_outliers(self, reading_time):
+        """
+        Given a list of times, calculates and removes outliers, which are the data points that
+        are outside the interquartile range of the data
+        :param reading_time: list, reading times for a specific question
+        :return: list, reading times for a specific question with outliers removed
+        """
+        reading_time.sort()
+        quartile_one = reading_time[math.trunc(len(reading_time) * 0.25)]
+        quartile_three = reading_time[math.trunc(len(reading_time) * 0.75)]
+        interquartile_range = quartile_three - quartile_one
+        lower_fence = quartile_one - (1.5 * interquartile_range)
+        upper_fence = quartile_three + (1.5 * interquartile_range)
+
+        view_time_two = 0
```

Comment:
Modifying this list in place has two problems:
- it's harder to read and figure out what's going on (more prone to error etc.)
- introduces side effects when this code could be purely functional (easier to reason about!)

Instead of going through and removing the elements that match the condition, construct a new list, into which you only append the elements that you want. The resulting code will be a lot cleaner and easier to manage.